src/lib.rs

```diff
@@ -1819,18 +1970,71 @@ where
     }
 }
 
-/// Are `bytes` aligned to `mem::align_of::<T>()`?
+trait AsAddress {
+    fn addr(self) -> usize;
+}
+
+impl<'a, T: ?Sized> AsAddress for &'a T {
+    #[inline(always)]
+    fn addr(self) -> usize {
+        #![allow(clippy::needless_return)]
+
+        let ptr: *const T = self;
+        // TODO(https://github.com/rust-lang/rust/issues/95228): Use `.addr()`
+        // instead of `as usize` once it's stable, and get rid of this `allow`.
+        // Currently, `as usize` is the only way to accomplish this.
+        #[allow(clippy::as_conversions)]
+        return ptr.cast::<()>() as usize;
+    }
+}
+
+impl<'a, T: ?Sized> AsAddress for &'a mut T {
+    #[inline(always)]
+    fn addr(self) -> usize {
+        #![allow(clippy::needless_return)]
+
+        let ptr: *mut T = self;
+        // TODO(https://github.com/rust-lang/rust/issues/95228): Use `.addr()`
+        // instead of `as usize` once it's stable, and get rid of this `allow`.
+        // Currently, `as usize` is the only way to accomplish this.
+        #[allow(clippy::as_conversions)]
+        return ptr.cast::<()>() as usize;
+    }
+}
+
+impl<T: ?Sized> AsAddress for *const T {
+    #[inline(always)]
+    fn addr(self) -> usize {
+        #![allow(clippy::needless_return)]
+
+        // TODO(https://github.com/rust-lang/rust/issues/95228): Use `.addr()`
+        // instead of `as usize` once it's stable, and get rid of this `allow`.
+        // Currently, `as usize` is the only way to accomplish this.
+        #[allow(clippy::as_conversions)]
+        return self.cast::<()>() as usize;
+    }
+}
+
+impl<T: ?Sized> AsAddress for *mut T {
+    #[inline(always)]
+    fn addr(self) -> usize {
+        #![allow(clippy::needless_return)]
+
+        // TODO(https://github.com/rust-lang/rust/issues/95228): Use `.addr()`
+        // instead of `as usize` once it's stable, and get rid of this `allow`.
+        // Currently, `as usize` is the only way to accomplish this.
+        #[allow(clippy::as_conversions)]
+        return self.cast::<()>() as usize;
+    }
+}
+
+/// Is `t` aligned to `mem::align_of::<U>()`?
 #[inline(always)]
-fn aligned_to<T>(bytes: &[u8]) -> bool {
-    // TODO(https://github.com/rust-lang/rust/issues/91126): Use `.to_bits()`
-    // instead of `as usize` once it's stable, and get rid of this `allow`.
-    // Currently, `as usize` is the only way to accomplish this.
-    #[allow(clippy::as_conversions)]
-    let addr = bytes.as_ptr().cast::<()>() as usize;
-    // `mem::align_of::<T>()` is guaranteed to return a non-zero value, which in
+fn aligned_to<T: AsAddress, U>(t: T) -> bool {
+    // `mem::align_of::<U>()` is guaranteed to return a non-zero value, which in
     // turn guarantees that this mod operation will not panic.
     #[allow(clippy::arithmetic_side_effects)]
-    let remainder = addr % mem::align_of::<T>();
+    let remainder = t.addr() % mem::align_of::<U>();
```

Comment:
NB: This can just be `t.addr() & (mem::align_of::<U>() - 1) == 0`. If you want to micro-optimize (you might not) then that might give slightly better codegen since we know `mem::align_of` always returns a power of two.